### DISCUSSION TITLE
bootstrap flux via ssh tunnel

### DISCUSSION THREAD

**Discussion by wibed** [fluxcd/flux2#4414] [General] [answered]
### the command i issue:
```
flux bootstrap git \
--ssh-hostname=192.168.1.241:2222 \
--url=ssh://git@127.0.0.1:2222/srv/bigbird.git \
--private-key-file=$HOME/.ssh/flux \
--path=./src/clusterA/staging
```

### the error i get:
```
► cloning branch "main" from Git repository "ssh://git@127.0.0.1:2222/srv/bigbird.git"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ component manifests are up to date
► installing components in "flux-system" namespace
✔ installed components
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
► generating source secret
✗ SSH key scan for host 192.168.1.241:2222 failed, error: dial tcp 192.168.1.241:2222: i/o timeout
```

### the logs tell me:
```
2023-11-19T05:23:43.415Z error GitRepository/flux-system.flux-system - Reconciler error failed to checkout and determine revision: unable to clone 'ssh://git@127.0.0.1:2222/srv/bigbird.git': dial tcp 127.0.0.1:2222: connect: connection refused
```

### what i try to do:
id like to bootstrap a remote cluster. only access i have is a ssh tunnel established as follows:
`ssh -L2222:10.0.48.69:2222 root@<REDACTED IP>`

### what i tried:
- i am able to establish a connection via ssh  from within the pod
```shell
# command
ssh git@192.168.1.241 -p 2222 -i /home/git/.ssh/gitsrv
# output
The authenticity of host '[192.168.1.241]:2222 ([192.168.1.241]:2222)' can't be established.
ED25519 key fingerprint is SHA256:1mX2/aZ3zzLdqkdAQiJb3D4uQtuna3vzR2ANFCVBS8s.
This key is not known by any other names.
Are you sure you want to continue connecting (yes/no/[fingerprint])? y
Please type 'yes', 'no' or the fingerprint: yes
Warning: Permanently added '[192.168.1.241]:2222' (ED25519) to the list of known hosts.
You've successfully authenticated, but I do not
provide interactive shell access.
Connection to 192.168.1.241 closed.
``` 

### addendum
it works if i am onsite. on the same network as the server. (my ip: 192.168.1.35; serverIP: 192.168.1.241)
i failed aswell exchanging `ssh-hostname` and `url` but i believe to have used the command correctly.

i am certain to use the command right. yet expect the ssh-hostname to replace the initial url hostname and url port once deployed.
which it does not

ps.: there is a nat between 192.168.1.0 and 10.0.48.0, the port 2222 is forwarded

**Comment by makkes** [comment]
Bootstrap doesn't provide a way for you to use a different clone URL locally than what is used on the cluster. So if your repository is only reachable through 127.0.0.1:2222 locally and only through 192.168.1.241:2222 from the cluster, bootstrap won't work for you.

By providing `--url=ssh://git@127.0.0.1:2222/srv/bigbird.git` you tell Flux's source-controller running on the cluster to establish a connection to 127.0.0.1 on port 2222 which obviously won't work. You'll have to provide a URL that is actually reachable from the source-controller Pod.

**Reply by wibed** [reply]
i am still confused about the lack of this obvious functionality.

on one hand the command does try to clone from the terminal it is executed from.
on the other it does expect the source to be available given the same address from the remote location.

within local networks and a ssh tunnel this is not possible except i do provide a custom dns on the cluster end and edit /etc/hosts on the terminal end.

how is it necessary for flux to share a common ancestor like the github repo? does the ip play any role given it is solely used for the bootstrap purpose?

**Reply by makkes** [reply]
If by "this obvious functionality" you mean the possibility to use a different repository URL locally and on the cluster then I would suggest that users haven't asked for this so far.

> on one hand the command does try to clone from the terminal it is executed from.
on the other it does expect the source to be available given the same address from the remote location.

Yes. Maybe a flag like `--local-url` would be beneficial in your case and we'd happily review a PR suggesting adding such a flag. We can discuss on the details about its impact there then.

**Reply by makkes** [reply, marked as answer]
As to your specific issue I suggest you don't use `flux bootstrap` but 
- `flux install --export ... > gotk-components.yaml` together with 
- `flux create source git --export ... > gotk-sync.yaml` and
- `flux create kustomization --export ... >> gotk-sync.yaml`

Then commit and push the generated manifests to your Git repository and `kubectl apply` them from your terminal. This will have the same effect as `flux bootstrap` but requires a little more manual work.

You may also need to pre-create the sync Secret using `flux create source git` in order for Flux to be able to authenticate towards the Git repository server.

Updating the Flux version is much easier then. You can take a look at [this guide](https://fluxcd.io/flux/installation/upgrade/) for details.

**Reply by wibed** [reply]
this wouldnt need a docker registry would it?
i am referring to: https://fluxcd.io/flux/installation/configuration/air-gapped/

edit:
havent seen the link at the bottom. but i have noticed that you export twice into `gotk-sync.yaml`

**Reply by makkes** [reply]
>  havent seen the link at the bottom. but i have noticed that you export twice into gotk-sync.yaml

Oops, should've been `>>`. Fixed that in the original comment.

**Reply by wibed** [reply]
just for completeness sake.


how do these differ:
```
flux install \
--components-extra image-reflector-controller,image-automation-controller \
--export > src/clusterA/staging/flux-system/gotk-components.yaml
```

```
flux install --export > src/clusterA/staging/flux-system/gotk-components.yaml
```

is different from 


edit:

i get it, it adds the controllers referred here:
https://fluxcd.io/flux/guides/image-update/

**Reply by makkes** [reply]
Yes, IRC and IAC aren't installed by default.

**Reply by wibed** [reply]
thank you i did follow these guidelines to follow yours.
https://github.com/fluxcd/flux2/discussions/1499

**Comment by wibed** [comment]
thank you i did follow these guidelines to follow yours.
https://github.com/fluxcd/flux2/discussions/1499
